Dynamic App list sorting (#1085)
smartdevicelink/sdl_evolution: proposals/NNNN-Dynamic-App-List-Sorting.md

 4. The apps are registered to the SDL Core by using the `RegisuterAppInterface`.
 5. The sorting order is performed according to the following procedures.
	1. Prioritize the `priority` parameter of `application_policies` in the policy table.
	2. Regarding to the apps that there is no priority value in the `priority` parameter, they are prioritized according to the sorting methods which are set by the `sorting_pattern` of the `app_list_sorting` in the policy table.
 6. The sorting list is sent to the HMI by using the `UpdateAppList`.

```json
app_list_sorting: {
    "sorting_pattern": "apphmitype",
    "apphmitype_sorting_list": {
        "NAVIGATION": 1,
        "MEDIA": 2,
        "MESSAGE": 3,
        "INFORMATION": 4,
        "COMMUNICATION": 5,
        "SOCIAL": 6,
        "PROJECTION": 7,
        "REMOTE_CONTROL": 8,
        "SYSTEM": 9,
        "DEFAULT": 10,
        "BACKGROUND_PROCESS": 11,
        "TESTING": 12
    }
}
```

> By changing the parameters of  `apphmitype_sorting_list` and `priority` created by each OEM in the policy table, they can display the SDL app icon according to each OEM's intended order.

Author: @theresalech -san,
Thank you for your review.
The word "They" indicates the meaning of  "changing the parameters of  `apphmitype_sorting_list` and `priority`".
Is the sentence "it can display the SDL app icon" correct?

<b>2. Clarification of icon display sequence</b><br>
Currently, there is no sequence provided that demonstrates how the icon is displayed. Figure 1 shows the sequence in which the process of this proposal is added to the existing sequence from the program code of SDL Core.
- The black text indicates the existing order described in the program code.
- The red text indicates the process added by this proposal.

<b>Figure 1.</b> SDL App Icon display sequence
![Figure1_SDL_App_Icon_display_sequence.PNG](../assets/proposals/NNNN-Dynamic-App-List-Sorting/Figure1_SDL_App_Icon_display_sequence.PNG)

## Potential downsides

> There is no potential downside since it is a modification to the current process of the SDL app icon display order.

Reviewer: Please provide some potential downsides in this section. That is, what are some downsides to modifying the current process?

Author: @theresalech -san,
Thank you for your review.
I modified the potencial downsides and made a commit it.
Please confirm it.
Thank you.

> As there is a large modification required to implement multiple sorting methods, the developer must do the implementation carefully and it will take a long time. Also, it may need to remove the sorting function from the HMI.

Reviewer: ```suggestion
As there is a large modification required to implement multiple sorting methods, the developer must do the implementation carefully and it will take a long time. Also, the system may need to remove the sorting function from the HMI.
```

## Impact on existing code

> The existing sorting process is left as it is. Furthermore, as processes/lists related to sorting are added, SDL Core is affected. Also, since the priority information of `apphmitype_sorting_list` is stored, the policy server is affected, too. However, I believe their changes are minor version changes because there are no changes in RPC and parameters.

Reviewer: ```suggestion
The existing sorting process is left as it is. Furthermore, as processes/lists related to sorting are added, SDL Core is affected. Also, since the priority information of `appHMIType_sorting_list` is stored, the policy server is affected, too. However, I believe their changes are minor version changes because there are no changes in RPC and parameters.
```

## Alternatives considered

> None.

Reviewer: Please provide alternatives in this section. This could include some of the concepts that were originally presented in SDL 0314.

Author: @theresalech -san,
Thank you for your review.
I modified to add the alternatives considered and made a commit it.
Please confirm it.
Thank you.

## Potential downsides

> Since it is necessary to implement the large amount of modification due to implement multiple sorting methods, the developer must do the implementation carefully and also, it will take a long time.

Author: @theresalech -san,
Thank you for your review as always.
I will modify to add the more information and make a commit.
Could you please confirm it if I made a commit.
Thank you.

> As there is a large modification required to implement multiple sorting methods, the developer must do the implementation carefully and it will take a long time. Also, it may need to remove the sorting function from the HMI.

Author: @theresalech -san,
It specifies the implementation of multiple sorting mthods.
Thank you.

> As there is a large modification required to implement multiple sorting methods, the developer must do the implementation carefully and it will take a long time. Also, it may need to remove the sorting function from the HMI.

Reviewer 2: @theresalech -san
I apologize for interrupting your busy work.
@Akihiro-Miyazaki was reply your question.
I would be grateful if you could comment on whether we can keep "it" as "it" or better to rewrite "it" as "System".

> As there is a large modification required to implement multiple sorting methods, the developer must do the implementation carefully and it will take a long time. Also, it may need to remove the sorting function from the HMI.

Reviewer: Hi @Shohei-Kawano, I've provided a suggested change in a separate comment.

## Impact on existing code

The existing sorting process is left as it is. Furthermore, as processes/lists related to sorting are added, SDL Core is affected. Also, since the priority information of `appHMIType_sorting_list` is stored, the policy server is affected, too. However, I believe their changes are minor version changes because there are no changes in RPC and parameters.

## Alternatives considered

> The alternative sorting method is user defined sort order. This sorting method would be a feature that some OEMs would want. However, since this specific sorting method would only be part of the HMI and have no other SDL component integration, we think it is difficult for SDL component to realize this function. Therefore, we think it is better this function to include into the HMI Integration Guidelines and SDL Core Guides documentation.

Reviewer: >we think it is difficult for SDL component to realize this function.

Can you please clarify what you mean by "realize this function"?

Author: @theresalech -san,
Thank you for your review.
What I mean by "realize this function" is to add the user defined sort order function to the SDL Core.
Thank you.

> The alternative sorting method is user defined sort order. This sorting method would be a feature that some OEMs would want. However, since this specific sorting method would only be part of the HMI and have no other SDL component integration, we think it is difficult for SDL component to realize this function. Therefore, we think it is better this function to include into the HMI Integration Guidelines and SDL Core Guides documentation.

Reviewer: >Therefore, we think it is better this function to include into the HMI Integration Guidelines and SDL Core Guides documentation.

Are you suggesting adding the alternative sorting method (user defined sort order) to the HMI Integration Guidelines and SDL Core Guides documentation in addition to your proposed solution of dynamic app list sorting? Or are you referencing the new dynamic app list sorting here, and saying that it should be documented in the HMI Integration Guidelines and SDL Core Guides?

Author: @theresalech -san,
Thank you for your review.
What I proposed as the alternative sorting method is to add the user defined sort order function.
So, I will remove the sentence "Therefore, we think it is better this function to include into the HMI Integration Guidelines and SDL Core Guides documentation." .
Thank you.

> The alternative sorting method is user defined sort order. This sorting method would be a feature that some OEMs would want. However, since this specific sorting method would only be part of the HMI and have no other SDL component integration, we think it is difficult for SDL component to realize the user defined sort order function.

Reviewer: ```suggestion
The alternative sorting method is user-defined sort order. This sorting method would be a feature that some OEMs would want. However, since this specific sorting method would only be part of the HMI and have no other SDL component integration, we think it is difficult for SDL Core to contain the user-defined sort order function.
```
I've suggested rewording the last sentence based on your response [here](https://github.com/smartdevicelink/sdl_evolution/pull/1085#discussion_r508153699). Please commit this suggested change if I'm understanding correctly, or let me know if I misunderstood and change as you see fit.
